ext/zlib/zlib.c

```diff
@@ -1088,6 +1042,68 @@ zstream_unblock_func(void *ptr)
     args->interrupt = 1;
 }
 
+static int
+zstream_run_func(struct zstream_run_args *args)
+{
+    struct zstream *z = args->z;
+    int state;
+    uInt n;
+
+    int err = Z_OK;
+    while (!args->interrupt) {
+        n = z->stream.avail_out;
+
+        rb_str_locktmp(z->buf);
+
+#ifndef RB_NOGVL_UBF_ASYNC_SAFE
+        err = (int)(VALUE)rb_thread_call_without_gvl(zstream_run_once, (void *)args, zstream_unblock_func, (void *)args);
+#else
+        err = (int)(VALUE)rb_nogvl(zstream_run_once, (void *)args, zstream_unblock_func, (void *)args, RB_NOGVL_UBF_ASYNC_SAFE);
+#endif
+
+        // Assuming that `rb_nogvl` doesn't raise an exception, we don't need to deal with `rb_ensure`.
```

Comment:
It could raise an exception. It checks for interrupt after acquiring the GVL, so things from `Thread#raise` and signals could make it raise.

You could pass `RB_NOGVL_INTR_FAIL` to nogvl, but then if you could still have a fiber scheduler that raises in `#blocking_region`.

Reply:
Ah okay so we will definitely need to use `rb_ensure` here then.

Reply:
Okay, I believe this is now fixed.